Datatable dropdown usability tweaks

bennothommo opened the pull request:
Adds a couple of usability tweaks to the dropdown cell type in the data table widget, to more closely mimic a native dropdown field.

1. Pressing the up or down arrow keys when the cell is focused but with the dropdown closed will select the previous or next item automatically and set it as the cell value. This does prevent the usual table function of going to the previous or next row when focused on a dropdown cell, but I think it's a worthwhile trade-off. When the dropdown is open, the up and down arrows work the same as previously implemented.

2. Typing out characters will initiate a search and select the first matching option automatically and set it as the cell value. For example, for the following options:
- Apples
- Oranges
- Bananas
Typing out `o` and `r` on the keyboard will automatically select the Oranges option.

LukeTowers commented:
Could you provide some gifs of this in action?

bennothommo commented:
Sorry for the delay - trying to work out video editing on Linux :P

Example of pressing the up and down arrow keys on the dropdown cells in a data table:
![Up Down Arrow Example](https://i.imgur.com/16XqrtY.gif)

Example of typing out the specific data type in the dropdown cells:
![Typing Example](https://i.imgur.com/gwgEqzJ.gif)

bennothommo commented:
@w20k Done :)

LukeTowers commented:
@w20k let me know when you're done testing this. @bennothommo looks great, this is something that's bugged me for a long time!

w20k commented:
@LukeTowers, will have to look through the commit and that should be it. Tomorrow, I guess 😉

w20k commented:
@bennothommo found few issues with the drop-down.

**Steps**:
- Open Dropdown, search for closest of the selected element and then move your mouse to select any
**Selectable Madness:**
_(Selected - Integer, searched for a Date. Pressed D-A, and mouseover to Timestamp)_
<img width="645" alt="screenshot 2018-12-15 at 13 30 29" src="https://user-images.githubusercontent.com/1053320/50042618-32313900-006e-11e9-9b4e-656d1bd594f5.png">

@bennothommo, one more thing, if you don't mind. Could you look at why you can not close drop-down if you select outside of the list? Thanks!

bennothommo commented:
@w20k Thanks heaps for the feedback. I've changed the behaviour of the dropdown somewhat - now only one item in a dropdown list will be focused at any one time, depending on the last input method used. For example, if you hover over the items, the hovered item will be "focused" and highlighted. If you then press the up or down arrow, or start typing in a search query, the focus will then shift away from the hovered item to the respective new item. This works similar to native dropdown menus.

Example:
![Dropdown new behaviour](https://i.imgur.com/nwdj6j4.gif)

With regards to the dropdown not hiding when clicked outside the cell, it does hide when you click another cell, but doesn't hide if you click the toolbar or the empty space underneath (even though both are a part of the "table" element). I've added an "unfocusTable" to the click event of the table widget if you click outside its cells, which should hopefully resolve this issue.

w20k commented:
Awesome! Thanks, @bennothommo!

@LukeTowers, Testing Status - ✅
